pallets/parachain-staking/src/inflation.rs

```diff
@@ -0,0 +1,253 @@
+
+#[cfg(feature = "std")]
+use serde::{Deserialize, Serialize};
+
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Eq, PartialEq, Clone, Encode, Decode, Default, RuntimeDebug)]
+pub struct RewardRate {
+	pub annual: Perquintill,
+	pub per_block: Perquintill,
+}
+
+/// Convert annual reward rate to per_block
+fn annual_to_per_block(rate: Perquintill) -> Perquintill {
+	rate / YEARS
+}
+
+impl RewardRate {
+	pub fn new(rate: Perquintill) -> Self {
+		RewardRate {
+			annual: rate,
+			per_block: annual_to_per_block(rate),
+		}
+	}
+}
+
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Eq, PartialEq, Clone, Encode, Decode, Default, RuntimeDebug)]
+pub struct StakingInfo {
+	/// Maximum staking rate
+	pub max_rate: Perquintill,
+	/// Reward rate
+	pub reward_rate: RewardRate,
+}
+
+impl StakingInfo {
+	pub fn new(max_rate: Perquintill, annual_reward_rate: Perquintill) -> Self {
+		StakingInfo {
+			max_rate,
+			reward_rate: RewardRate::new(annual_reward_rate),
+		}
+	}
+
+	/// Calculate newly minted rewards on coinbase, e.g.,
+	/// reward = rewards_per_block * staking_rate
+	pub fn compute_block_rewards<T: Config>(&self, stake: BalanceOf<T>, total_issuance: BalanceOf<T>) -> BalanceOf<T> {
+		let staking_rate = Perquintill::from_rational(stake, total_issuance).min(self.max_rate);
+		// multiplication with perbill cannot overflow
+		let rewards = self.reward_rate.per_block * total_issuance;
+		// multiplication with perbill cannot overflow
+		staking_rate * rewards
+	}
+}
+
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Eq, PartialEq, Clone, Encode, Decode, Default, RuntimeDebug)]
+pub struct InflationInfo {
+	pub collator: StakingInfo,
+	pub delegator: StakingInfo,
+}
+
+impl InflationInfo {
+	pub fn new(
+		collator_max_rate_percentage: Perquintill,
+		collator_annual_reward_rate_percentage: Perquintill,
+		delegator_max_rate_percentage: Perquintill,
+		delegator_annual_reward_rate_percentage: Perquintill,
+	) -> Self {
+		Self {
+			collator: StakingInfo::new(collator_max_rate_percentage, collator_annual_reward_rate_percentage),
+			delegator: StakingInfo::new(delegator_max_rate_percentage, delegator_annual_reward_rate_percentage),
+		}
+	}
+
+	/// Compute coinbase rewards for collators and delegators based on the
+	/// current staking rate and the InflationInfo
+	pub fn block_issuance<T: Config>(
+		&self,
+		collator_stake: BalanceOf<T>,
+		delegator_stake: BalanceOf<T>,
+	) -> (BalanceOf<T>, BalanceOf<T>) {
+		let circulating = T::Currency::total_issuance();
+
+		let collator_rewards = self.collator.compute_block_rewards::<T>(collator_stake, circulating);
+		let delegator_rewards = self.delegator.compute_block_rewards::<T>(delegator_stake, circulating);
+
+		(collator_rewards, delegator_rewards)
+	}
+
+	/// Check whether the annual reward rate is approx. the per_block reward
+	/// rate multiplied with the number of blocks per year
+	pub fn is_valid(&self) -> bool {
```

Comment:
Should we maybe have this checked every time an `InflationInfo` instance is created, rather than having to check each time that the instance is valid? So we would be sure that each time `InflationInfo` is used, it is valid.